timers/timers_test.go

```diff
@@ -0,0 +1,411 @@
+
+	bs.Schedule(uint64(0), fakeWork)
+	assertQueuedTimers(t, bs, 0, 1)
+
+	bs.Schedule(uint64(1), fakeWork)
+	assertQueuedTimers(t, bs, 1, 1)
+
+	bs.Schedule(uint64(7), fakeWork)
+	assertQueuedTimers(t, bs, 7, 1)
+
+	bs.Schedule(uint64(8), fakeWork)
+	assertQueuedTimers(t, bs, 0, 2)
+}
+
+func TestBucketListGatherExpired(t *testing.T) {
+	bs := newBucketList()
+	for i := range bs.buckets {
+		bs.Schedule(uint64(i), fakeWork)
+	}
+	assert.Equal(t, 0, bs.GatherExpired(0, 0, 0).len(), "Unexpected no. expired timers in range [0,0).")
+	assert.Equal(t, 3, bs.GatherExpired(0, 3, 3).len(), "Unexpected no. expired timers in range [0,3).")
+	assert.Equal(t, 0, bs.GatherExpired(0, 3, 3).len(), "Unexpected no. expired timers in range [0,3) on re-gather.")
+	assert.Equal(t, 5, bs.GatherExpired(3, 100, 100).len(), "Unexpected no. expired timers in range [3,100).")
+
+	// We should be leaving unexpired timers in place.
+	bs.Schedule(8, fakeWork)
+	bs.Schedule(8, fakeWork)
+	assert.Equal(t, 0, bs.GatherExpired(0, 1, 1).len(), "Unexpected no. expired timers in range [0,1).")
+	assertQueuedTimers(t, bs, 0, 2)
+}
+
+func TestBucketListClear(t *testing.T) {
+	bs := newBucketList()
+	for i := range bs.buckets {
+		bs.Schedule(uint64(i), fakeWork)
+	}
+	bs.Schedule(uint64(1<<20), fakeWork)
+	all := bs.Clear()
+	assert.Equal(t, 9, all.len(), "Unexpected number of timers after clearing bucketList.")
+	for i := range bs.buckets {
+		assert.Panics(t, func() { bs.buckets[i].Unlock() }, "Expected all buckets to be unlocked.")
+	}
+}
+
+func TestBucketListClearIgnoresCanceledTimers(t *testing.T) {
+	bs := newBucketList()
+	for i := range bs.buckets {
+		bs.Schedule(uint64(i), fakeWork).Stop()
+	}
+	all := bs.Clear()
+	assert.Equal(t, 0, all.len(), "Expected all timers to be canceled.")
+}
+
+func TestTimerBucketCalculations(t *testing.T) {
+	tests := []struct {
+		tick, max time.Duration
+		buckets   int
+	}{
+		// If everything is a power of two, buckets = max / tick.
+		{2, 8, 4},
+		// Tick gets bumped up to 1<<3, so we need 3 buckets to support 20ns
+		// max without overlap. We then bump that to 1<<2.
+		{6, 20, 4},
+	}
+
+	for _, tt := range tests {
+		w := NewWheel(tt.tick, tt.max)
+		defer w.Stop()
+		assert.Equal(t, tt.buckets, len(w.buckets.buckets), "Unexpected number of buckets.")
+	}
+}
+
+func TestTimersScheduleAndCancel(t *testing.T) {
+	const (
+		numTimers    = 20
+		tickDuration = 1 << 22
+		maxDelay     = tickDuration * numTimers
+	)
+	w, c := fakeWheel(tickDuration)
+	defer w.Stop()
+
+	var wg sync.WaitGroup
+	scheduled := make([]*Timer, 0, numTimers)
+	for i := time.Duration(0); i < numTimers; i++ {
+		scheduled = append(scheduled, w.AfterFunc(i*tickDuration, wg.Done))
+		// wg.Done() panics if the counter is less than zero, but these extra
+		// calls should never fire.
+		canceled := w.AfterFunc(i*tickDuration, wg.Done)
+		assert.True(t, canceled.Stop())
+	}
+
+	for elapsed := time.Duration(0); elapsed < maxDelay; elapsed += tickDuration {
```

Comment:
since you never use `elapsed` and `maxDelay` in the body of the loop, why not just make this simpler:
```go
for i := 0; i < maxTimers; i++ {
  wg.Add(1)
  c.Add(tickDuration)
  wg.Wait()
}
```